sphinx/util/osutil.py

```diff
@@ -193,3 +194,76 @@ def cd(target_dir):
         yield
     finally:
         os.chdir(cwd)
+
+
+class FileAvoidWrite(object):
+    """File-like object that buffers output and only writes if content changed.
+
+    Use this class like when writing to a file to avoid touching the original
+    file if the content hasn't changed. This is useful in scenarios where file
+    mtime is used to invalidate caches or trigger new behavior.
+
+    When writing to this file handle, all writes are buffered until the object
+    is closed.
+
+    Objects can be used as context managers.
+    """
+    def __init__(self, path):
+        self._path = path
+        self._io = None
+
+    def write(self, data):
+        if not self._io:
+            if isinstance(data, text_type):
+                self._io = StringIO()
+            else:
+                self._io = BytesIO()
+
+        self._io.write(data)
+
+    def close(self):
+        """Stop accepting writes and write file, if needed."""
+        if not self._io:
+            raise Exception('FileAvoidWrite does not support empty files.')
+
+        buf = self.getvalue()
+        self._io.close()
+
+        r_mode = 'r'
+        w_mode = 'w'
+        if isinstance(self._io, BytesIO):
+            r_mode = 'rb'
+            w_mode = 'wb'
+
+        old_content = None
+
+        try:
+            old_f = open(self._path, r_mode)
```

Comment:
Nested try-except is a bit ugly in this case. It would be better if you use `with` statement.